File: src/main/java/frc/team5190/robot/drive/DTRSubsystem.java
package frc.team5190.robot.drive;

import com.ctre.phoenix.motorcontrol.ControlMode;
import com.ctre.phoenix.motorcontrol.FeedbackDevice;
import com.ctre.phoenix.motorcontrol.can.TalonSRX;
import com.kauailabs.navx.frc.AHRS;
import edu.wpi.first.wpilibj.GenericHID;
import edu.wpi.first.wpilibj.SPI;
import edu.wpi.first.wpilibj.command.Subsystem;
import frc.team5190.robot.Robot;
import frc.team5190.robot.util.Constants;

@SuppressWarnings("WeakerAccess")
public class DTRSubsystem extends Subsystem {

    public TalonSRX frontLeft;
    public TalonSRX frontRight;
    public TalonSRX rearLeft;
    public TalonSRX rearRight;

    public AHRS navX;

    public DTRSubsystem() {
        frontLeft = new TalonSRX(Constants.FRONT_LEFT);
        frontRight = new TalonSRX(Constants.FRONT_RIGHT);
        rearLeft = new TalonSRX(Constants.REAR_LEFT);
        rearRight = new TalonSRX(Constants.REAR_RIGHT);

        DTRHelper.configurePIDF(frontLeft, 0.6, 0, 0, 1, Constants.HIGH_GEAR_MAX, Constants.WHEEL_RADIUS, Constants.TICKS_PER_ROTATION, FeedbackDevice.QuadEncoder);
        DTRHelper.configurePIDF(frontRight, 0.6, 0, 0, 1, Constants.HIGH_GEAR_MAX, Constants.WHEEL_RADIUS, Constants.TICKS_PER_ROTATION, FeedbackDevice.QuadEncoder);

        rearLeft.follow(frontLeft);
        rearRight.follow(frontRight);

        navX = new AHRS(SPI.Port.kMXP);
    }

    @Override
    protected void initDefaultCommand() {
        this.setDefaultCommand(new DTRCommand());
    }

    public void xboxDrive() {
        tankDrive(-Robot.oi.getXbox().getY(GenericHID.Hand.kLeft), Robot.oi.getXbox().getY(GenericHID.Hand.kRight), ControlMode.Velocity);
    }

    public void tankDrive(double leftValue, double rightValue, ControlMode mode) {
        if (leftValue >= 0.0) {
            leftValue = leftValue * leftValue;
        } else {
            leftValue = -(leftValue * leftValue);
        }
        if (rightValue >= 0.0) {
            rightValue = rightValue * rightValue;
        } else {
            rightValue = -(rightValue * rightValue);
        }

        if (mode == ControlMode.Velocity) {
            leftValue *= 1049;
            rightValue *= 1049;
        }

        frontLeft.set(mode, leftValue);
        frontRight.set(mode, rightValue);
    }

    public void reset() {
        navX.reset();
        navX.setAngleAdjustment(90);
        tankDrive(0, 0, ControlMode.PercentOutput);
    }
}

File: src/main/kotlin/frc/team5190/robot/drive/DTRSubsystem.kt
package frc.team5190.robot.drive

import com.ctre.phoenix.motorcontrol.ControlMode
import com.ctre.phoenix.motorcontrol.FeedbackDevice
import com.ctre.phoenix.motorcontrol.StatusFrame
import com.ctre.phoenix.motorcontrol.StatusFrameEnhanced
import com.ctre.phoenix.motorcontrol.can.TalonSRX
import com.kauailabs.navx.frc.AHRS
import edu.wpi.first.wpilibj.SPI
import edu.wpi.first.wpilibj.command.Subsystem
import frc.team5190.robot.OI
import frc.team5190.robot.util.Hardware
import frc.team5190.robot.util.MotorIDs

object DriveTrain : Subsystem() {

    val frontLeft = TalonSRX(MotorIDs.FRONT_LEFT)
    val frontRight = TalonSRX(MotorIDs.FRONT_RIGHT)

    private val rearLeft = TalonSRX(MotorIDs.REAR_LEFT)
    private val rearRight = TalonSRX(MotorIDs.REAR_RIGHT)

    private val navX = AHRS(SPI.Port.kMXP)

    init {

        DTRHelper.configurePIDF(frontLeft, 0.0, 0.0, 0.0, 1.0, rpm = Hardware.MAX_RPM.toDouble(),
                sensorUnitsPerRotation = Hardware.NATIVE_UNITS_PER_ROTATION.toDouble(), dev = FeedbackDevice.QuadEncoder)

        DTRHelper.configurePIDF(frontRight, 0.0, 0.0, 0.0, 1.0, rpm = Hardware.MAX_RPM.toDouble(),
                sensorUnitsPerRotation = Hardware.NATIVE_UNITS_PER_ROTATION.toDouble(), dev = FeedbackDevice.QuadEncoder)

        frontLeft.configMotionProfileTrajectoryPeriod(10, 10)
        frontRight.configMotionProfileTrajectoryPeriod(10, 10)

        frontLeft.setStatusFramePeriod(StatusFrameEnhanced.Status_10_MotionMagic, 50, 10)
        frontRight.setStatusFramePeriod(StatusFrame.Status_10_MotionMagic, 10, 10)

        frontLeft.inverted = true
        rearLeft.inverted = true

        frontLeft.configNeutralDeadband(0.04, 10)
        frontRight.configNeutralDeadband(0.04, 10)

        rearLeft.follow(frontLeft)
        rearRight.follow(frontRight)
    }

    fun driveWithXbox() {
        this.tankDrive(OI.getLeftY(), OI.getRightY(), ControlMode.Velocity)
    }

    fun tankDrive(leftOutput: Double, rightOutput: Double, mode: ControlMode) {
        when (mode) {
            ControlMode.PercentOutput -> {
                frontLeft.set(mode, leftOutput)
                frontRight.set(mode, rightOutput)

                println("Left: " + frontLeft.sensorCollection.quadraturePosition + ", Right: " + frontRight.sensorCollection.quadraturePosition)
            }
            ControlMode.Velocity -> {
                frontLeft.set(mode, leftOutput * Hardware.MAX_NATIVE_UNITS_PER_100_MS)
                frontRight.set(mode, rightOutput * Hardware.MAX_NATIVE_UNITS_PER_100_MS)
            }
            else -> {
                println("Learn to dab sir.")
            }
        }
    }

    override fun initDefaultCommand() {
        this.defaultCommand = DriveCommand()
    }

    fun reset() {
        frontLeft.set(ControlMode.PercentOutput, 0.0)
        frontRight.set(ControlMode.PercentOutput, 0.0)
        navX.angleAdjustment = 90.0
    }
}